#### C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [1]:
# Import necessary libraries
import numpy as np  # For numerical operations
import pandas as pd  # For data manipulation

from sklearn.metrics import mean_squared_error  # To calculate the Mean Squared Error
from sklearn.model_selection import train_test_split  # To split the dataset into training and test sets
from sklearn.preprocessing import StandardScaler  # To normalize the data
from tensorflow.keras.layers import Dense  # To define the layers of the neural network
from tensorflow.keras.models import Sequential  # To build the neural network
from tensorflow.keras.optimizers import Adam  # Optimizer for training the model

In [2]:
# Load the dataset
url = "concrete_data.csv"  # Path to the dataset
data = pd.read_csv(url)  # Load the dataset into a pandas DataFrame

In [3]:
# Split data into predictors (X) and target variable (y)
X = data.drop("Strength", axis=1)  # Features/predictors (all columns except "Strength")
y = data["Strength"]  # Target variable ("Strength")

In [7]:
# Part A: Build Baseline Model
def baseline_model(X, y):
    """
    Builds and evaluates a baseline regression model using Keras.
    The model:
    - Has one hidden layer with 10 nodes and ReLU activation.
    - Uses the Adam optimizer and mean squared error loss function.
    The process is repeated 50 times, and the mean and standard deviation of the MSEs are computed.

    Parameters:
    X: Features (predictors)
    y: Target variable (concrete strength)

    Returns:
    Mean and standard deviation of the MSEs from 50 iterations.
    """
    mse_list = []  # List to store MSEs from each iteration

    # Repeat the process 50 times
    for _ in range(50):
        # Split the dataset into training (70%) and testing (30%) sets
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=np.random.randint(0, 100))
        
        # Build the neural network model
        model = Sequential([
            Dense(10, activation='relu', input_shape=(X_train.shape[1],)),  # Hidden layer with 10 nodes
            Dense(1)  # Output layer with a single node (for regression)
        ])
        
        # Compile the model using Adam optimizer and mean squared error loss
        model.compile(optimizer=Adam(), loss='mean_squared_error')
        
        # Train the model on the training data for 50 epochs
        model.fit(X_train, y_train, epochs=50, verbose=0)
        
        # Evaluate the model on the test data
        y_pred = model.predict(X_test, verbose=0)  # Predict on the test set
        mse = mean_squared_error(y_test, y_pred)  # Calculate Mean Squared Error
        mse_list.append(mse)  # Append the MSE to the list
    
    # Return the mean and standard deviation of the MSEs
    return np.mean(mse_list), np.std(mse_list)

# Part B: Normalize the data
def normalized_model(X, y):
    """
    Builds and evaluates a regression model using normalized data.
    The normalization process:
    - Subtracts the mean and divides by the standard deviation for each feature.
    The rest of the process is the same as the baseline model, with 50 iterations.

    Parameters:
    X: Features (predictors)
    y: Target variable (concrete strength)

    Returns:
    Mean and standard deviation of the MSEs from 50 iterations with normalized data.
    """
    # Normalize the data using StandardScaler
    scaler = StandardScaler()  # Initialize the scaler
    X_normalized = scaler.fit_transform(X)  # Fit the scaler to X and transform it
    
    # Call the baseline model function with the normalized data
    return baseline_model(X_normalized, y)

# Part C: Increase the number of epochs
def increased_epochs_model(X, y):
    """
    Builds and evaluates a regression model with an increased number of epochs (100).
    The model:
    - Has one hidden layer with 10 nodes and ReLU activation.
    - Uses the Adam optimizer and mean squared error loss function.
    The dataset is normalized before training, and the process is repeated 50 times.

    Parameters:
    X: Features (predictors)
    y: Target variable (concrete strength)

    Returns:
    Mean and standard deviation of the MSEs from 50 iterations with increased epochs.
    """
    mse_list = []  # List to store MSEs from each iteration

    # Normalize the dataset to zero mean and unit variance
    scaler = StandardScaler()  # Initialize the StandardScaler
    X_normalized = scaler.fit_transform(X)  # Fit and transform X

    # Repeat the process 50 times for robust evaluation
    for _ in range(50):
        # Split the normalized dataset into training and testing sets (70% train, 30% test)
        X_train, X_test, y_train, y_test = train_test_split(
            X_normalized, y, test_size=0.3, random_state=np.random.randint(0, 100)
        )
        
        # Build the neural network model
        model = Sequential([
            Dense(10, activation='relu', input_shape=(X_train.shape[1],)),  # Hidden layer with 10 nodes
            Dense(1)  # Output layer with a single node (for regression task)
        ])
        
        # Compile the model using Adam optimizer and mean squared error loss function
        model.compile(optimizer=Adam(), loss='mean_squared_error')
        
        # Train the model on the training data for 100 epochs
        # Increasing the number of epochs allows the model to train longer, potentially improving accuracy
        model.fit(X_train, y_train, epochs=100, verbose=0)
        
        # Evaluate the model on the test data
        y_pred = model.predict(X_test, verbose=0)  # Predict on the test set
        mse = mean_squared_error(y_test, y_pred)  # Calculate the Mean Squared Error
        mse_list.append(mse)  # Append the MSE to the list
    
    # Return the mean and standard deviation of the MSEs
    return np.mean(mse_list), np.std(mse_list)

In [5]:
# Call the function to run the model with increased epochs and capture the results
mean_c, std_c = increased_epochs_model(X, y)

In [6]:
print("Part C - Increased Epochs (100): Mean MSE =", mean_c, "Std MSE =", std_c)

Part C - Increased Epochs (100): Mean MSE = 170.7973410637508 Std MSE = 19.64843488987875


#### How does the mean of the mean squared errors compare to that from Step B?

The comparison between the Mean MSE values from Step B (Normalized Model) and Step C (Increased Epochs) shows the following:

1. Mean MSE Comparison:
    - Step B (Normalized Model): Mean MSE = 369.76
    - Step C (Increased Epochs): Mean MSE = 170.80

The mean MSE in Step C is significantly lower compared to Step B. This indicates that increasing the number of epochs (from 50 to 100) allowed the model to train for a longer period, enabling it to learn the data's underlying patterns more effectively, thereby improving prediction accuracy.

2. Standard Deviation Comparison:
    - Step B (Normalized Model): Std MSE = 93.85
    - Step C (Increased Epochs): Std MSE = 19.65

The standard deviation of MSE in Step C is also considerably lower than in Step B. This reduction in variability suggests that increasing the number of epochs not only improved accuracy but also made the model's performance more consistent across the 50 iterations.

#### Conclusion:
Increasing the number of epochs from 50 (Step B) to 100 (Step C) resulted in a substantial improvement in the model's accuracy (lower mean MSE) and stability (lower standard deviation of MSE). This shows that the model benefited from the additional training time to better capture the relationships within the data.